src/server/config/__tests__/complete.js

```diff
@@ -1,105 +1,165 @@
+    });
   });
 
-  it('should transform server.ssl.cert to server.ssl.certificate', function () {
-    const kbnServer = {
-      settings: {
-        server: {
-          ssl: {
-            cert: 'path/to/cert'
+  describe('all settings used', () => {
+    it('should not throw', function () {
+      const { callCompleteMixin } = setup({
+        settings: {
+          used: true
+        },
+        configValues: {
+          used: true
+        },
+      });
+
+      callCompleteMixin();
+    });
+
+    describe('more config values than settings', () => {
+      it('should not throw', function () {
+        const { callCompleteMixin } = setup({
+          settings: {
+            used: true
+          },
+          configValues: {
+            used: true,
+            foo: 'bar'
           }
-        }
-      }
-    };
+        });
 
-    const server = {
-      decorate: noop,
-      log: sinon.spy()
-    };
+        callCompleteMixin();
+      });
+    });
+  });
 
-    const config = {
-      get: sinon.stub().returns({
-        server: {
-          ssl: {
-            certificate: 'path/to/cert'
+  describe('some settings unused', () => {
+    it('should throw an error', function () {
+      const { callCompleteMixin } = setup({
+        settings: {
+          unused: true
+        },
+        configValues: {
+          used: true
+        }
+      });
+
+      expect(callCompleteMixin).to.throwError('"unused" not applied');
+    });
+
+    describe('error thrown', () => {
+      it('has correct code, processExitCode, and message', () => {
+        const { callCompleteMixin } = setup({
+          settings: {
+            unused: true,
+            foo: 'bar',
+            namespace: {
+              with: {
+                sub: {
+                  keys: true
+                }
+              }
+            }
           }
+        });
+
+        expect(callCompleteMixin).to.throwError((error) => {
+          expect(error).to.have.property('code', 'InvalidConfig');
+          expect(error).to.have.property('processExitCode', 64);
+          expect(error.message).to.contain('"unused", "foo", and "namespace.with.sub.keys"');
+        });
+      });
+    });
+  });
+
+  describe('deprecation support', () => {
+    it('should transform settings when determining what is unused', function () {
+      sandbox.spy(transformDeprecationsNS, 'transformDeprecations');
+
+      const settings = {
+        foo: 1
+      };
+
+      const { callCompleteMixin } = setup({
+        settings,
+        configValues: {
+          ...settings
         }
-      })
-    };
+      });
+
+      callCompleteMixin();
+      sinon.assert.calledOnce(transformDeprecations);
+      expect(transformDeprecations.firstCall.args[0]).to.be(settings);
+    });
+
+    it('should use transformed settings when considering what is used', function () {
+      sandbox.stub(transformDeprecationsNS, 'transformDeprecations', (settings) => {
+        settings.bar = settings.foo;
```

Comment:
And the reason that doesn't throw an error of some sort if because we're using `_.keys()` rather than `Object.keys()`... I'm going to switch to `Object.keys()` so the error isn't swallowed like this